### PR TITLE
Add adapter option disallowing foreign keys

### PR DESCRIPTION
**Pull request by promulo**
### Summary

Sometimes, apps can choose not to make use of foreign keys even though the underlying database supports it. One reason, for example, could be having trouble migrating big tables online due to the way [some tools](https://github.com/soundcloud/lhm/issues/34) handle foreign key constraints.

Regardless of the motivation for not wanting to use foreign key constraints, the most likely way of doing that would be to patch `ActiveRecord::ConnectionAdapters::AbstractAdapter#support_foreign_keys` to return `false` and override the concrete adapters' implementation, which will depend on individual database systems.

This PR introduces a new adapter config option in Active Record which allows apps to override the connection adapters' implementation if they need/want to. It can be added to `database.yml` like the following:

```yaml
development:
  <<: *default
  database: db/development.sqlite3
  foreign_keys: false
```

Thanks for the feedback!

### CODE REVIEW

File: railties/lib/rails/application/configuration.rb [hunk 1]
```diff
@@ -283,6 +283,10 @@ def load_defaults(target_version)
           if respond_to?(:action_controller)
             action_controller.allow_deprecated_parameters_hash_equality = false
           end
+
+          if respond_to?(:active_record)
+            active_record.use_foreign_keys = true
+          end
```
Comment (etiennebarrie) [inline railties/lib/rails/application/configuration.rb:289]: Do we need this given `true` is the default?

Comment (promulo) [reply railties/lib/rails/application/configuration.rb:289]: We actually do, otherwise the config value would be `nil` when referenced like this:

```
irb(main):001:0> Rails.application.config.active_record.use_foreign_keys
=> nil
```

**Review by skipkayhil — COMMENTED** [review]
Make sure you document this in https://github.com/rails/rails/blob/main/guides/source/configuring.md#configuring-active-record as well

**Comment by yahonda** [comment]
Thanks for opening a pull request. I am against this change I understand there are some use cases where no foreign keys are preferred. However, this kind of requirement should be done by migration files, not introducing another global setting.

```ruby
  def change
    create_table :authors do |t|
      t.string :name
      t.timestamps
    end

    create_table :books do |t|
      t.belongs_to :author, foreign_key: false
      t.datetime :published_at
      t.timestamps
    end
```

**Comment by eileencodes** [comment]
Hey @promulo! We can't change the name of `supports_foreign_keys?`. It is a public API and those can't be changed or renamed without a deprecation. Either this would need to be a separate setting or the name needs to be maintained. If we do add a new config option I'd prefer something more explicit like `disallow_foreign_keys`.

> Thanks for opening a pull request. I am against this change I understand there are some use cases where no foreign keys are preferred. However, this kind of requirement should be done by migration files, not introducing another global setting.

@yahonda while the migration does handle this fine it requires every engineer writing a migration to know that all migrations need this. The goal here is to turn off foreign keys for an entire application without engineers needing to know they have to do that every time. 

A compromise could be adding an option to the configuration hash in the database.yml. Then it can be set for an entire DB without setting for an entire app. I like this option the best because some apps might not want to turn them off entirely and leave it up to the database configuration. In this case I think the option name would be `foreign_keys` and default to true.

```
development:
  database: my_db
  foreign_keys: false
```

File: railties/lib/rails/application/configuration.rb [hunk 2]
```diff
@@ -283,6 +283,10 @@ def load_defaults(target_version)
           if respond_to?(:action_controller)
             action_controller.allow_deprecated_parameters_hash_equality = false
           end
+
+          if respond_to?(:active_record)
+            active_record.use_foreign_keys = true
```
Comment (casperisfine) [inline railties/lib/rails/application/configuration.rb:288]: Might make more sense to have this as an adapter configuration, no?

E.g we'd pass `use_foreign_key: false` in the `database.yml`, and that would cause `supports_foreign_keys?` to return `false`? That wouldn't change any API, and allow to configure this on a per DB basis.

**Comment by simi** [comment]
Even I understand your app needs @eileencodes, as @yahonda mentioned, this could lead to inconsistent `schema.rb` (or `structure.sql`) if I understand it well and that makes me :-1: on this change (even done in `database.yml`) as well.

On our side we do similar kind of DB sanity checks in our test suite. Other idea would be to introduce custom DB adapter and overload those methods in there. :thinking:

**Comment by byroot** [comment]
> this could lead to inconsistent `schema.rb` (or `structure.sql` )

I really don't see how this can be a concern. Presumably, if you take the decision to no longer use foreign key checks, you'd cleanup the existing tables.

>  Other idea would be to introduce custom DB adapter and overload those methods in there.

Yeah, but for a single method... it really doesn't hurt to add it as a default adapter option. Let's be practical.

**Comment by yahonda** [comment]
Thanks for the explanation. Setting `foreign_keys` in database.yml looks good to me .

```ruby
development:
  database: my_db
  foreign_keys: false
```

**Comment by simi** [comment]
> > this could lead to inconsistent `schema.rb` (or `structure.sql` )
> 
> I really don't see how this can be a concern. Presumably, if you take the decision to no longer use foreign key checks, you'd cleanup the existing tables.

I was thinking about case when this is switched in the middle of development and `schema.rb` is used. Loading `schema.rb` into DB (for example by `db:test:prepare` task) will result in different structure locally (when created by `schema.rb`) and different structure remotely (where managed by migrations). The problem with keeping this in `database.yml` makes it simple to have different values per environment as well (since in production often `DATABASE_URL` is used). To make it safe on this side maybe `ActiveRecord::InternalMetadata` could be utilised to freeze this option and ensure it is not changed on the fly.

> > Other idea would be to introduce custom DB adapter and overload those methods in there.
> 
> Yeah, but for a single method... it really doesn't hurt to add it as a default adapter option. Let's be practical.

Currently it seems to me it is based on special requirement of one app. That could be practical for that one app, but not for the rails itself. I was thinking about this again and the same result could be also achieved using custom Rubocop rule. There is a lot of alternatives which makes me thinking why this needs to be part of Rails itself.

The other more general (and thus potentially more useful for all apps) approach would be to be able to configure what your adapter supports at the time of initialising it (and make it possible to specify in `database.yml` manually as well). That way any `supports_*_?` method would be possible to be customised for app's purpose. That will most likely need refactoring of `supports_*_?` methods and will be complex patch :(.

---

Anyway if it will be decided to stick with this simple approach, would it be possible to at least add comment somewhere about potentially inconsistent DB schema?

**Comment by byroot** [comment]
> The problem with keeping this in database.yml makes it simple to have different values per environment as well 

You currently can have different adapters per environments, that will cause way more issues than that. Our users are grown-ups.

> Currently it seems to me it is based on special requirement of one app.

Absolutely not, there's way more than one app that use `LHM`, `gh-ost` or `pt-online-schema-change` (the later can deal with foreign key, but that makes it slower etc). Basically anyone with a large MySQL DB has this problem.

> could be also achieved using custom Rubocop rule

Using linters doesn't provide as good of an experience. 

> if it will be decided to stick with this simple approach

It's already decided, I'm just waiting for @promulo to update.

> would it be possible to at least add comment somewhere about potentially inconsistent DB schema?

Sure, that param will be documented, there can be a quick note.

**Comment by simi** [comment]
> > would it be possible to at least add comment somewhere about potentially inconsistent DB schema?
> 
> Sure, that param will be documented, there can be a quick note.

At least we agree on something. :smile:  Thanks for the patience and all the info @byroot.

**Comment by simi** [comment]
:thinking: Couldn't https://github.com/rails/rails/pull/45324 be used for the same purpose as well? It is mentioned in the summary. That seems a more general approach to me.

> Disallow certain operations; their strategy could raise if that method is called

**Comment by promulo** [comment]
Thanks everybody @byroot, @eileencodes, @yahonda, @simi, @skipkayhil for the reviews and input. 🙏🏽 

Just a quick clarification, this approach would not remove/change any existing APIs. `AbstractAdapter#support_foreign_keys` would still exist and be used by `use_foreign_keys?` in `SchemaStatements`.

I am fine with changing this to be an option in `database.yml`. This PR will be updated soon.

**Comment by promulo** [comment]
Sorry for the delay in getting back to this. I've update the branch.

There are a couple tests failing which pass locally and are not really related to my changes (at least not directly).

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 3]
```diff
@@ -1565,6 +1569,10 @@ def extract_foreign_key_action(specifier)
           end
         end
 
+        def foreign_keys_enabled?
+          @config[:foreign_keys].nil? || @config[:foreign_keys]
```
Comment (etiennebarrie) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1573]: Maybe this? 
```suggestion
          @config.fetch(:foreign_keys, true)
```

Comment (promulo) [reply activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:1573]: Much better indeed, thanks!

File: activerecord/CHANGELOG.md [hunk 4]
```diff
@@ -1,37 +1,52 @@
...
+    *Paulo Barros*
+
 *   Add configurable deprecation warning for singular associations
 
     This adds a deprecation warning when using the plural name of a singular associations in `where`.
     It is possible to opt into the new more performant behavior with `config.active_record.allow_deprecated_singular_associations_name = false`
 
     *Adam Hess*
 
-*   Run transactional callbacks on the freshest instance to save a given
```
Comment (yahonda) [inline activerecord/CHANGELOG.md:8]: Would you restore the original one, not to remove the training spaces in this pull request ? These changes are not related to this pull request goal. You can open another one if necessary.

**Review by yahonda — CHANGES_REQUESTED** [review]

File: activerecord/CHANGELOG.md (hunk 4 above)
Comment (yahonda) [reply activerecord/CHANGELOG.md:8]: I should have mentioned that removing trailing spaces may be considered as cosmetic changes.

**Review by yahonda — APPROVED** [review]

**Comment by byroot** [comment]
Unrelated failure on 3.1, but LGTM.
